docs/DesignDocs/CustomAttestation_V3.md

```diff
@@ -0,0 +1,374 @@
+#define OE_EVIDENCE_FLAGS_REMOTE_ATTESTATION 0x00000001
+#endif
+```
+
+### New verifier callback function
+
+```C
+/**
+ * The verifier attestion role. The verifier is reponsible for verifying the
+ * attestation evidence and must implement the functions below.
+ */
+typedef struct _oe_verifier oe_verifier_t;
+struct _oe_verifier
+{
+    ...
+
+    /**
+     * get_claims
+     *
+     * Get claims needed to generate evidence.
+     *
+     * @param[out] claims The list of claims.
+     * @param[out] claims_length The length of the claims list.
+     * @retval OE_OK on success.
+     * @retval other appropriate error code.
+     */
+    oe_result_t (*get_claims(
+        oe_claim_t** claims,
+        size_t* claims_length);
+};
+```
+
+## Code samples
+More plugin code snippets are availabe in [CustomAttestation.md](https://github.com/openenclave/openenclave/blob/master/docs/DesignDocs/CustomAttestation.md). The sample below is mainly for demonstrating the use of new APIs.
+
+The plugin-ware verifier application, which can either be the enclave or the host, can verify the evidence like this:
+
+`verifier.c`
+
+```C
+#include <sgx/verifier.h>
+
+/* Register SGX DCAP plugin upon initialization. */
+oe_register_verifier(oe_sgx_plugin_dcap_verifier(), NULL, 0);
+
+/* Get registered verifier format IDs. */
+oe_get_registered_verifier_format_ids(&format_ids, &format_ids_length);
+
+/* Send the format IDs that can be verified to the attester.
+   Protocol is up to enclave and verifier. */
+send(VERIFIER_SOCKET_FD, format_ids, format_ids_length*sizeof(oe_uuid_t), 0);
+
+/* Receive a selected format ID. */
+recv(ENCLAVE_SOCKET_FD, &format_id, &format_id_size, 0);
+
+/* Get claims for the format ID. */
+oe_get_verifier_claims(format_id, &verifier_claims, &verifier_claims_length);
+
+/* Serialize the claims and send the claims buffer to the attester */
+_serialize_claims(verifier_claims, verifier_claims_length, &claims_buffer, &claims_buffer_size);
+send(VERIFIER_SOCKET_FD, claims_buffer, claims_buffer_size, 0);
+
+/* Receive evidence and endorsement buffer from enclave. */
+recv(ENCLAVE_SOCKET_FD, evidence, evidence_size, 0);
+recv(ENCLAVE_SOCKET_FD, endorsements, endorsements_size, 0);
+
+/* Verify evidence. Can check the claims if desired. */
+oe_verify_evidence(
+    evidence,
+    evidence_size,
+    endorsements,
+    endorsements_size,
+    NULL,
+    0,
+    &claims,
+    &claims_size);
+
+/* Free data and unregister plugin. */
+oe_free_claims_list(verifier_claims, verifier_claims_length);
+oe_free_format_ids(format_ids);
+oe_free_claims_list(claims, claims_size);
+oe_unregister_verifier(oe_sgx_plugin_dcap_verifier());
+```
+
+The plugin-aware application enclave can generate the SGX evidence using the plugin like this:
+
+`attester.c`
+
+```C
+#include <sgx/attester.h>
+
+/* Register SGX plugins upon initialization. */
+oe_register_attester(oe_sgx_plugin_dcap_attester(), NULL, 0);
+oe_register_attester(oe_sgx_plugin_local_attester(), NULL, 0);
```

Comment:
As described in my previous comments, using SGX specific generic initialization functions should be more flexible.

- e_result_t oe_sgx_get_attester_plugins(oe_attester_t** attesters, size_t attesters_length)
- oe_result_t oe_sgx_get_verifier_plugins(oe_verifier_t** verifiers, size_t verifiers_length)

As compared to implementation-specific functions like oe_sgx_plugin_local_attester(), oe_sgx_plugin_dcap_attester(), or oe_sgx_plugin_dcap_verifier().

Reply:
oe_sgx_get_attesters and oe_sgx_get_verifiers added.

---
In reply to: [390025445](https://github.com/openenclave/openenclave/pull/2621#discussion_r390025445) [](ancestors = 390025445)